Add MASWE-0047, MASWE-0048, MASWE-0049, MASWE-0050, MASWE-0051, MASWE-0052 (#2919)
OWASP/owasp-mastg: weaknesses/MASVS-NETWORK/MASWE-0047.md

---

## Overview

> Certificate pinning is a security technique used to ensure that an app only trusts specific certificates or public keys when establishing secure connections. Insecure identity pinning occurs when the implementation of certificate or public key pinning is flawed or improperly configured. This weakness can leave the app vulnerable to Man-in-the-Middle (MITM) attacks and other security threats. Common issues include outdated pins, improper validation, accepting all certificates, or using insecure methods for dynamic pinning.

Reviewer: Should we rename weakness to use Certificate pinning? If not, this sentence should start with Identity pinning and quickly explain that it's also called certificate pinning or TLS pinning. This should be done in both cases even, to make sure all names are mentioned at the start.

Reviewer: Feels weird to have 'accepting all certificates' in here. We could go for 'insecure fallback' or something, or 'insecure default', but 'accepting all certificates' seems like the opposite of identity pinning and should be a separate weakness ("Not validating TLS certificate" or something)

Author: - Let's say Machine-in-the-Middle (MITM), see [NIST](https://csrc.nist.gov/glossary/term/man_in_the_middle_attack).
- I think we should make this more compact and link to the content we already have in https://mas.owasp.org/MASTG/0x04f-Testing-Network-Communication/#restricting-trust-identity-pinning 
- I noticed that we have mentions to the Modes of Introduction here so I removed it and instead I'm adding more content about why it's bad not to have it (what can go wrong) and the **limitations** (even if you have it, it's not foolproof).

```suggestion
[Identity pinning (aka. certificate pinning, public key pinning or TLS pinning)](../../Document/0x04f-Testing-Network-Communication/#restricting-trust-identity-pinning) refers to associating a mobile app with a specific cryptographic identity, such as a certificate or public key to ensure that the app only communicates with trusted servers. 

When a mobile app does not implement certificate pinning, or if it is implemented incorrectly, the app remains vulnerable to Machine-in-the-Middle (MITM) attacks which enable attackers to intercept and modify the communication between the app and the intended server. This occurs because when the app is presented a fraudulent certificate that the app may unknowingly trust, thereby gaining access to sensitive data or injecting malicious content into the data stream.

**Limitations**: Certificate pinning adds a layer of trust verification by ensuring that the app only accepts connections to servers with specific, pre-determined certificates or public keys. This reduces the risk of unauthorized interception, even if a trusted Certificate Authority (CA) is compromised. However, it is not foolproof:

- Attackers who can reverse-engineer the app may analyze and remove or modify the pre-defined pins or the certificate pinning logic to bypass the checks permanently.
- Attackers who can perform @MASTG-TECH-0051 techniques to manipulate the app and bypass pinning checks.

This highlights the importance of implementing certificate pinning **alongside other security measures** to enhance the app's resistance to advanced threats.
```

## Impact

- **Data Interception**: Sensitive data such as credentials, personal information, or financial details can be captured by unauthorized parties.
- **Data Manipulation**: Attackers can not only intercept but also manipulate data if pinning is not properly enforced.
- **Denial of Service**: Incorrect pinning may cause legitimate connections to fail, leading to service disruptions for users. For example, if a pinned certificate expires and is not updated, the app may be unable to establish secure connections.

## Modes of Introduction

- **Improper Configuration of Pinning Libraries**: Misconfiguring libraries like TrustKit, OkHttp's `CertificatePinner`, Volley, or AFNetworking's `SSLPinningMode`, leading to ineffective pinning.
- **Dynamic Pinning without Security**: Retrieving pins dynamically over insecure channels without proper validation, making it easy for attackers to supply malicious pins.

> - **Pinning to Insecure Certificates**: Pinning to self-signed, expired, or untrusted certificates that can be exploited.

Author: I see your point, maybe it makes sense to remove it from here since we have all the other weaknesses that cover the real problem which is not so much the pinning but the lack of certificate validation in the first place.

```suggestion
```

> - **Partial or Inconsistent Pinning**: Only applying pinning to certain network requests or failing to enforce it consistently across the app.

Author: Agreed

```suggestion
```

- **Weak Validation Logic**: Custom pinning implementations that do not correctly validate the certificate chain or public key. For example, accepting any certificate that chains to a trusted root instead of a specific certificate or public key.
- **Lack of Backup Pins**: Not including backup pins to prevent connectivity issues if the primary pin is no longer valid.

## Mitigations

- **Prefer Platform-provided Solutions**: Use platform-provided mechanisms like Android's Network Security Configuration (NSC) or iOS's App Transport Security (ATS) to enforce pinning.
- **Use Trusted Pinning Libraries**: Refrain from writing custom pinning logic; instead, rely on established and well-maintained libraries and frameworks (e.g., TrustKit, OkHttp's `CertificatePinner`) and ensure they are correctly configured according to best practices.
- **Secure Dynamic Pinning**: If dynamic pinning is necessary, retrieve pins over secure channels and validate them thoroughly before use.

> - **Pin to Public Keys Instead of Certificates**: Pin to the server's public keys rather than certificates to avoid issues with certificate expiration and renewals.

Reviewer 2: ```suggestion
- **Pin to Public Keys Instead of Certificates**: Pin to the certificates public key rather than the whole certificate to avoid issues with certificate expiration and renewals.
```

> - **Consistent Enforcement**: Apply pinning uniformly to all relevant network connections within the app.

Reviewer: ```suggestion
- **Consistent Enforcement**: Apply pinning uniformly to all network connections where you control the backend certificates.
```
